File: OJ/leet332/leet332.py
# leet332

from typing import List

from collections import deque 

# 这种「一笔画」问题与欧拉图或者半欧拉图有着紧密的联系，下面给出定义：
# 通过图中所有边恰好一次且行遍所有顶点的通路称为欧拉通路。
# 通过图中所有边恰好一次且行遍所有顶点的回路称为欧拉回路。
# 具有欧拉回路的无向图称为欧拉图。
# 具有欧拉通路但不具有欧拉回路的无向图称为半欧拉图。
# 
# 因为本题保证至少存在一种合理的路径，也就告诉了我们，这张图是一个欧拉图或者半欧拉图。我们只需要输出这条欧拉通路的路径即可。
# 如果没有保证至少存在一种合理的路径，我们需要判别这张图是否是欧拉图或者半欧拉图，具体地：
# 对于无向图 G，G 是欧拉图当且仅当 G 是连通的且没有奇度顶点。
# 对于无向图 G，G 是半欧拉图当且仅当 G 是连通的且 G 中恰有 2 个奇度顶点。
# 对于有向图 G，G 是欧拉图当且仅当 G 的所有顶点属于同一个强连通分量且每个顶点的入度和出度相同。
# 对于有向图 G，G 是半欧拉图当且仅当 G 的所有顶点属于同一个强连通分量且
# 恰有一个顶点的出度与入度差为 1；
# 恰有一个顶点的入度与出度差为 1；
# 所有其他顶点的入度和出度相同。


class Solution1:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        M = {v[0]:[] for v in tickets} 
        for v in tickets:
            M[v[0]].append(v[1])
        for k, v in M.items():
            M[k] = deque(sorted(v, reverse=True))
        
        start = "JFK"
        ans = [start]

        def recursive(start: str):
            if len(ans) == len(tickets) + 1: return True
            if start not in M: return False
            n_next = len(M[start])
            for _ in range(n_next):
                to = M[start].pop()
                ans.append(to)
                if recursive(to):
                    return True 
                ans.pop()
                M[start].appendleft(to)
            return False
        
        recursive(start)
        return ans

from collections import defaultdict
import heapq 


class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:

        def dfs(curr: str):
            while M[curr]:
                tmp = heapq.heappop(M[curr])
                dfs(tmp)
            stack.append(curr)

        M = defaultdict(list)
        
        for start, to in tickets:
            M[start].append(to)
        for k, v in M.items():
            heapq.heapify(v)

        stack = []

        dfs("JFK")

        return stack[::-1]

print(Solution().findItinerary([["MUC", "LHR"], ["JFK", "MUC"], ["SFO", "SJC"], ["LHR", "SFO"]]))
print(Solution().findItinerary([["JFK","SFO"],["JFK","ATL"],["SFO","ATL"],["ATL","JFK"],["ATL","SFO"]]))
print(Solution().findItinerary([["JFK","KUL"],["JFK","NRT"],["NRT","JFK"]]))

File: OJ/leet332/leet332.cpp
/*
leet332


*/


#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <tuple>
#include <cassert>
#include <utility>
using namespace std;

template<class T>
ostream& operator<<(ostream& os, const vector<T>& v) {
    for (const auto& vi: v) os << vi << ", "; os << endl;
    return os;
}

template<class T>
ostream& operator<<(ostream& os, const vector<vector<T>>& v) {
    for (const auto& row: v) os << row;
    return os;
}


class Solution1 {
public:
    vector<string> ans;
    unordered_map<string, deque<string>> M;
    int n;

    bool recursive(string& start) {
        // 朴素的递归解法
        if (ans.size() == n) return true;
        auto p = M.find(start);
        if (p == M.end()) return false;

        auto n_next = p->second.size();
        while (n_next--) {
            auto to = M[start].back();
            M[start].pop_back();
            ans.push_back(to);

            if (recursive(to)) { return true; }

            ans.pop_back();
            M[start].push_front(to);
        }
        return false;
    }
    vector<string> findItinerary(const vector<vector<string>>& tickets) {
        M.clear(); n = tickets.size() + 1;
        for (const auto& v: tickets) {
            M[v[0]].push_back(v[1]);
        }
        for (auto& [k, v]: M) {
            sort(v.begin(), v.end(), greater<string>());
        }
        
        string start = "JFK";
        ans = {start};

        recursive(start);

        return ans;
    }
};

class Solution2 {
public:
    unordered_map<string, priority_queue<string, vector<string>, std::greater<string>>> M;

    vector<string> stk;

    void dfs(const string& curr) {
        // 看成欧拉路径的求解问题
        // 只需注意到如果起点的入度与出度不同（差1），则终点唯一且必定是 dfs 第一个完成回溯的点
        // 因此对普通邻结点，按照从小到大的顺序访问即可
        // 终结点找到并入栈后，之后这些邻结点会按照从大到小的顺序入栈。因此逆序输出即可
        while (M.count(curr) && M[curr].size() > 0) {
            string tmp = M[curr].top();
            M[curr].pop();
            dfs(move(tmp));
        }
        stk.emplace_back(curr);
    }

    vector<string> findItinerary(const vector<vector<string>>& tickets) {
        for (auto& it : tickets) {
            M[it[0]].emplace(it[1]);
        }
        dfs("JFK");

        reverse(stk.begin(), stk.end());
        return stk;
    }
};

class Solution {
public:
    unordered_map<string, vector<string>> M;

    vector<string> stk;

    void dfs(const string& curr) {
        // 看成欧拉路径的求解问题
        // 只需注意到如果起点的入度与出度不同（差1），则终点唯一且必定是 dfs 第一个完成回溯的点
        // 因此对普通邻结点，按照从小到大的顺序访问即可
        // 终结点找到并入栈后，之后这些邻结点会按照从大到小的顺序入栈。因此逆序输出即可
        while (M.count(curr) && M[curr].size() > 0) {
            string tmp = M[curr].back();
            M[curr].pop_back();
            dfs(move(tmp));
        }
        stk.emplace_back(curr);
    }

    vector<string> findItinerary(const vector<vector<string>>& tickets) {
        for (auto& v : tickets) {
            M[v[0]].emplace_back(v[1]);
        }
        
        for (auto& [k, v]: M) {
            sort(v.begin(), v.end(), greater<string>());
        }
        dfs("JFK");

        return vector<string>(stk.rbegin(), stk.rend());
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;   
    // cout << sol.findItinerary( {{"MUC", "LHR"}, {"JFK", "MUC"}, {"SFO", "SJC"}, {"LHR", "SFO"}} );
    // cout << sol.findItinerary( {{"JFK","SFO"},{"JFK","ATL"},{"SFO","ATL"},{"ATL","JFK"},{"ATL","SFO"}} );
    cout << sol.findItinerary( {{"JFK","KUL"},{"JFK","NRT"},{"NRT","JFK"}, {"JFK","ABC"},{"ABC","JFK"}} );
    return 0;
}
